sympy/vector/functions.py

```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
     scalar : SymPy Expr
         The scalar field to compute the gradient of
 
-    coord_sys : CoordSysCartesian
-        The coordinate system to calculate the gradient in
-
     Examples
     ========
 
-    >>> from sympy.vector import CoordSysCartesian, gradient
-    >>> R = CoordSysCartesian('R')
+    >>> from sympy.vector import CoordSystem3D, gradient
+    >>> R = CoordSystem3D('R')
     >>> s1 = R.x*R.y*R.z
-    >>> gradient(s1, R)
+    >>> gradient(s1)
     R.y*R.z*R.i + R.x*R.z*R.j + R.x*R.y*R.k
     >>> s2 = 5*R.x**2*R.z
-    >>> gradient(s2, R)
+    >>> gradient(s2)
     10*R.x*R.z*R.i + 5*R.x**2*R.k
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```

Comment:
> Don't you feel that we should implement this before we continue?

@jbbskinny try to play around with `sympy.diffgeom`, especially the way how coordinate systems are connected there:

http://docs.sympy.org/dev/modules/diffgeom.html#sympy.diffgeom.CoordSystem

Consider just one think: all parameters passed to the class constructor have to be set on class initialization. Once an object has been created, its fields should not be modified. This is an important rule of SymPy, which unfortunately may restrict API choices.

This is how `.connect_to( )` is implemented:
https://github.com/sympy/sympy/blob/b6c26ae535323ff041864b47d2fcac743a763fbe/sympy/diffgeom/diffgeom.py#L215

I wouldn't go that way, but it's still worth comparing. I think the API should be flexible, some users may only desire to input the Lamé parameters, others may wish to specify the full transformation from another coordinate system. I believe we need an API which allows to do both, and I would play with the constructor arguments.

By the way... do you know there is a Wiki of SymPy? I'd suggest to draft some API change proposals before starting coding:
https://github.com/sympy/sympy/wiki